### PR TITLE
Add ability to toggle posts as pinned and unpinned [Fixes #1096]

### LINKED ISSUE #1096: Create the ability to toggle posts as pinned or unpinned
**Is your feature request related to a problem? Please describe.**

1. The Talawa home screen shows a newsfeed of  posts from Talawa users.
1. In the Talawa Admin console there are plans to have the ability for administrators to not only add posts to the news feed, but to also pin them to the top of the newsfeed
1. In anticipation of that feature we need the API to support this requirement
1. Look at the video below to see how the pinned posts are expected to be rendered on the Talawa mobile app screen

**Describe the solution you'd like**
1. Create ways to:
    1. store the pinned characteristic to every post
    1. update this characteristic to have one of two states, pinned or unpinned
    1. determine whether a post is pinned or not for the mobile app.
    1. create a way for the mobile app to get the N th. currently pinned post 
        1. always in reverse chronological order
        1. indicating whether the post is the last one in the list
1. 100% test code coverage 
1.  Update these Talawa-Docs  as necessary (This will require another issue created in that repo by you)
     1. [Schemas Page](https://docs.talawa.io/docs/developers/talawa-api/schemas) 
     1. [Mutations Page](https://docs.talawa.io/docs/developers/talawa-api/mutations) 
     1. [The markdown versions of those files can be found here](https://github.com/PalisadoesFoundation/talawa-docs/tree/develop/docs/developers/talawa-api). 

**Describe alternatives you've considered**
N/A

**Approach to be followed (optional)**
See above

**Additional context**
[IMAGE ALT TEXT HERE](https://www.youtube.com/watch?v=ix3eSReJj2k?t=79)

### PR DESCRIPTION
**Pull request by EshaanAgg**
**What kind of change does this PR introduce?**
Adds feature to implement marking posts as pinned and unpinned 

**Issue Number:**

Fixes #1096 

**Did you add tests for your changes?**
Yes

**Snapshots/Videos:**
NA

**If relevant, did you update the documentation?**
Will update the documentation after I get a sanity check for the mutations and queries implemented.

**Summary**
The PR introduces the following changes:
1. Add a `pinned` field on each post which is a boolean
2. Add two mutations, `pinPost` and `unpinPost`, each of which take in the `id` of the post as an argument and then sets the status of `pinned` property of the post to be `true` and `false` respectively.
3. Implemented the `pinnedPosts` query which returns all the posts which are pinned, latest first.
4. Implemented the `pinnedPostsByOrganization`  query, which returns all the pinned posts for a particular orginzation, latest first.

**Does this PR introduce a breaking change?**
NO

**Other information**
Currently we have a `postsByOrganizationConnection` query defined in our schema too. Do we need a corresponding query for `pinnedPosts` too?

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes

### CODE REVIEW

**Review by noman2002 — APPROVED** [review]
Looks good to me.

**Comment by Kevoniat** [comment]
looks good to me

**Comment by xoldyckk** [comment]
@palisadoes Don't merge yet. Wait for the discussion to conclude.
https://github.com/PalisadoesFoundation/talawa-api/issues/1096#issuecomment-1448421491

**Review by xoldyckk — COMMENTED** [review]
@EshaanAgg Take a look at this discussion. Some things can be improved.
https://github.com/PalisadoesFoundation/talawa-api/issues/1096

**Comment by xoldyckk** [comment]
@EshaanAgg You can make the changes as discussed. Also, just include pagination for pinnedPosts by default, it's no big deal.

**Comment by EshaanAgg** [comment]
Got it @xoldyckk.

File: src/resolvers/Mutation/createPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/createPost.ts:98]: Make this `args.data.pinned === true`. We don't need to check all other cases.

File: src/resolvers/Mutation/createPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/createPost.ts:100]: You're querying database two times, just to get current user's data. You've already done it before with currentUserExists. Instead of just using it to check whether user exists, fetch the user object, check its existence there and then also make use of its data here for comparision.

File: src/resolvers/Mutation/createPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/createPost.ts:104]: This comparision is hard to read. Simplify it to this:-
```
const currentUserIsOrganizationAdmin = currentUser.adminFor.some(organzationId=> organizationId.toString() === args.data.organizationId))

if (currentUser.userType !== "SUPERADMIN" && !currentUserIsOrganizationAdmin){
  throw error.....
}
```

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/togglePinnedPost.ts:71]: Change `p` to `postId`. Try to be explicit about variable names. Short names just create confusion in the long run.

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/togglePinnedPost.ts:87]: If you don't want the returned result for a database mutation you should not use `findOneAndUpdate`, just use `updateOne`.

Same goes for using `lean()`. It's only used to transform a mongoose document to plain old javascript object for returning as result. Not needed here.

**Comment by xoldyckk** [comment]
@EshaanAgg I've thought about the user experience some more and how the post will be available on the client side. If the information whether a post is pinned or not needs to be available on each post instead of just on a particular page which only lists pinnedPosts, then using this approach of putting pinned posts ids in organization will make it slow to query for state of pinned boolean for each individual post. Since we'll have to create a seperate resolver for each post which just checks whether the postId exists in `pinnedPosts` of the corresponding organization, it'll be very slow.

If admins can pin posts from any place, whether admin dashboard or the mobile app itself, then the `pinned` information needs to be present within the post model itself to reduce database queries. This user experience far outweighs the cost of a stale `pinned` field on the post object. So, I guess we should just put the `pinned` boolean inside the post object.

**Comment by EshaanAgg** [comment]
> @EshaanAgg I've thought about the user experience some more and how the post will be available on the client side. If the information whether a post is pinned or not needs to be available on each post instead of just on a particular page which only lists pinnedPosts, then using this approach of putting pinned posts ids in organization will make it slow to query for state of pinned boolean for each individual post. Since we'll have to create a seperate resolver for Post type named `pinned` which will just check whether the parent.id(postId) exists in `pinnedPosts` of the corresponding organization(parent.organization). And it'll have to be done for each post, it'll be very slow if someone is querying for a list of posts instead of just one post.
> 
> If admins can pin posts from any place, whether admin dashboard or the mobile app itself, then the `pinned` information needs to be present within the post object itself to reduce database queries. This user experience far outweighs the cost of a stale `pinned` field on the post object. So, I guess we should just put the `pinned` boolean inside the post object.

@xoldyckk I don't think so that we should be doing the same. The reason for the same is as on the mobile app, when the organizations' feed is loaded, it would make sense to make just one query to the organisation, and not make an individual query to filter out all the `pinned` posts from all the posts of the organization. Once the list of posts is loaded, the UI can pass that data down to each of the child post detail pages.

We shoud probably employ both these strategies, that is having pinnedPosts stored on the organisation model, as well as store the pinned status on each post individually, so that we can have performant rendering of the UI of pinned posts in the mobile app with a single request (no need to make a separate query for pinnedPosts, and orgDetails separately), and then use the pinned status on each post for the other concern of user experience as you explained.

**Comment by xoldyckk** [comment]
> > @EshaanAgg I've thought about the user experience some more and how the post will be available on the client side. If the information whether a post is pinned or not needs to be available on each post instead of just on a particular page which only lists pinnedPosts, then using this approach of putting pinned posts ids in organization will make it slow to query for state of pinned boolean for each individual post. Since we'll have to create a seperate resolver for Post type named `pinned` which will just check whether the parent.id(postId) exists in `pinnedPosts` of the corresponding organization(parent.organization). And it'll have to be done for each post, it'll be very slow if someone is querying for a list of posts instead of just one post.
> > If admins can pin posts from any place, whether admin dashboard or the mobile app itself, then the `pinned` information needs to be present within the post object itself to reduce database queries. This user experience far outweighs the cost of a stale `pinned` field on the post object. So, I guess we should just put the `pinned` boolean inside the post object.
> 
> @xoldyckk I don't think so that we should be doing the same. The reason for the same is as on the mobile app, when the organizations' feed is loaded, it would make sense to make just one query to the organisation to fetch all of its details as well as the pinned posts, and not make an individual query to filter out all the `pinned` posts from all the posts of the organization. Once the list of posts is loaded, the UI can pass that data down to each of the child post detail pages.
> 
> We shoud probably employ both these strategies, that is having pinnedPosts stored on the organisation model, as well as store the pinned status on each post individually, so that we can have performant rendering of the UI of pinned posts in the mobile app with a single request (no need to make a separate query for pinnedPosts, and orgDetails separately), and then use the pinned status on each post for the other concern of user experience as you explained.
> 
> What do you think?

Yes, the second option is a bit better yes. But not a whole lot of performance gains.

Also, you're probably thinking that filtering organization posts based on `pinned` boolean will be slow, but it isn't the case. Since a boolean can only have two possible values sorting posts according to `pinned` field will be very fast. Databases are very smartly designed to handle these things.

Also, we should just index the organizationId field on post object. That way filtering the posts according to different organizations will be wayyy faster. It will not only be beneficial for `pinned` posts but also for querying posts for organizations by their foreign keys, whether the posts are `pinned` or not.

As a last resort, you can index the `pinned` field as well to achieve more speed but it'll not make much difference because it's just a boolean and can only have two distinct values at any time.

**Comment by EshaanAgg** [comment]
Yes I agree that the gain in performance wouldn't be much, but atleast it would be less burden and more flexibility for clients. 
Thanks for the detailed discussion on the performance aspect, and I'll read about creating an index on the `organizationID` for the `Post` model so that they can be queried more optimally. 

I'll also update the PR to follow the joint approach, and resolve all the conversations that you have provided above.

File: src/typeDefs/types.ts
Comment (EshaanAgg) [reply src/typeDefs/types.ts:202]: Got it.

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/togglePinnedPost.ts:46]: Ahh makes sense. An absent minded mistake on my end. Will rectify the same.

File: src/resolvers/Mutation/createPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/createPost.ts:98]: The current implementation and one you are suggesting are equivalent. With your approach, we would only get the true value if `true === true`, which is a redundant check, JS will treat `null`, `undefined` and `false` all to be `falsy` by deafult.

File: src/resolvers/Mutation/createPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/createPost.ts:100]: Got it.

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/togglePinnedPost.ts:52]: Got it. Will do.

File: src/resolvers/Mutation/createPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/createPost.ts:104]: Got it.

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/togglePinnedPost.ts:71]: Got it.

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/togglePinnedPost.ts:87]: Okay got it.

**Comment by xoldyckk** [comment]
> Yes I agree that the gain in performance wouldn't be much, but atleast it would be less burden and more flexibility for clients. Thanks for the detailed discussion on the performance aspect, and I'll read about creating an index on the `organizationID` for the `Post` model so that they can be queried more optimally.
> 
> I'll also update the PR to follow the joint approach, and resolve all the conversations that you have provided above.

Flexibility for clients is expressed in graphql schema not database. How you efficiently resolve the data for the flexibility graphql is offering is the backend's concern.

**Comment by EshaanAgg** [comment]
Getting your concerns @xoldyckk! Data duplication is as a general rule should be avoided, but here for our specific use case, I think we do have a bit of use for it. Probably not the most efficient implementation, but a good one nonetheless given our specific use case.

**Comment by xoldyckk** [comment]
> Getting your concerns @xoldyckk! Data duplication is as a general rule should be avoided, but here for our specific use case, I think we do have a bit of use for it. Probably not the most efficient implementation, but a good one nonetheless given our specific use case.

OK

File: src/resolvers/Mutation/createPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/createPost.ts:127]: Arrange the checks in a way that `pinned` is set in the same step the post is created.

File: src/resolvers/Mutation/createPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/createPost.ts:105]: I don't think you should throw error when you've already created the post. Do it before creating the post. 

Or don't throw the error and instead, create the post but change the value of input `pinned` boolean to false, if the user who sent it is a normal user.

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/togglePinnedPost.ts:72]: Since findOneAndUpdate returns the updated post, return it where you're updating it. Creating updatePost variable is useless here.

`return await Post.findOneAndUpdate....`

File: src/resolvers/Mutation/createPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/createPost.ts:105]: In the latest push, I am checking to see if the user is authorized to create the same or not. If not, Ill throw the error, otherwise I continue to make the post after that.

File: src/resolvers/Mutation/createPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/createPost.ts:110]: What if the `pinned` variable is null, undefined? We should not allow those values in the database. Or even it they're not allowed, this should not result in resolver failing to create a post. Check how mongodb/mongoose deals with it and whether we need to transform the input before storing it.

File: tests/resolvers/Mutation/togglePinnedPost.spec.ts
Comment (xoldyckk) [inline tests/resolvers/Mutation/togglePinnedPost.spec.ts:180]: You also have to check that on the actual post, the boolean is toggled.

File: src/resolvers/Mutation/createPost.ts
Comment (EshaanAgg) [reply src/resolvers/Mutation/createPost.ts:110]: Used a check in the latest commit to only restrict it to true/false

File: tests/resolvers/Mutation/togglePinnedPost.spec.ts
Comment (EshaanAgg) [reply tests/resolvers/Mutation/togglePinnedPost.spec.ts:183]: Updated the tests

**Comment by EshaanAgg** [comment]
Got it @xoldyckk! Superadmin should also have access to do so right?

**Comment by xoldyckk** [comment]
> Got it @xoldyckk! Superadmin should also have access to do so right?

IDK tbh.

**Comment by palisadoes** [comment]
> Got it @xoldyckk! Superadmin should also have access to do so right?

Yes. Super Admins should be able to do whatever an Admin can do for their organization

**Comment by EshaanAgg** [comment]
Yes @palisadoes, will do so as soon as all the changes in this PR are approved. Would make a new PR in `talawa-docs` and add reference to the original issue in the same.

File: src/resolvers/Mutation/togglePinnedPost.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/togglePinnedPost.ts:1]: rename the function to `togglePostPin`. Everything else looks good.

Comment (EshaanAgg) [reply src/resolvers/Mutation/togglePinnedPost.ts]: Done.

**Review by xoldyckk — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]

**Pull request merged**
